Blog - gain insights in the Databaricks Jobs with Arcus Background Jobs v0.2 (#105)
arcus-azure/arcus: community/blog/2020-12 Gain Insights in the Databricks Jobs with Arcus Background Jobs/README.md

# Gain insights in the Databricks Jobs with Arcus Background Jobs

> Starting from [Arcus BackgroundJobs v0.2](https://github.com/arcus-azure/arcus.backgroundjobs/releases/tag/v0.2.0) we have introduced a new library to monitor Databricks clusters. The problem with these clusters was that there's no job metrics out-of-the-box provided so the end result of the job (success/failure) is swallowed. People would have to look up the status of the Databricks job themselves to figure out how it ran.

Reviewer: ```suggestion
Starting from [Arcus Background Jobs v0.2](https://github.com/arcus-azure/arcus.backgroundjobs/releases/tag/v0.2.0) we have introduced a new library to monitor Databricks clusters. The challenge with Databricks clusters in Azure is that they do not provide out-of-the-box job metrics in Azure Monitor which makes it hard to operate them. People would have to look up the status of the Databricks job themselves to figure out how it ran and using automated alerts is hard.
```

> This post will show you how our new version of the **Arcus BackgroundJobs** library will help you with this problem so the next time you run a Databricks job, you'll be notified right away!

Reviewer: ```suggestion
This post will show you how our new version of the **Arcus Background Jobs** library will help you with this problem so the next time you run a Databricks job, you'll be notified right away!
```

## Cluster jobs

> Here's a take on some Databricks jobs which have been run on a cluster. Note the end result is shown in the last collumn.

Reviewer: ```suggestion
Here's a take on some Databricks jobs which have been run on a cluster. Note the end result is shown in the last column.
```

The end result of this post would be that we could notify the consumer somehow of this outcome without them looking it up in the Databricks cluster itself. 

![Cluster jobs](./images/cluster-jobs.png)

## Azure Function Databricks Job Metrics

> The **Arcus BackgroundJobs** library includes a new package in the v0.2 version called [Arcus.BackgroundJobs.Databricks](https://www.nuget.org/packages/Arcus.BackgroundJobs.Databricks/). This package contains all the necessary capabilities to query the Databricks cluster so we can retrieve the end result of the completed jobs. 

Reviewer: ```suggestion
The **Arcus Background Jobs** library includes a new package in the v0.2 version called [Arcus.BackgroundJobs.Databricks](https://www.nuget.org/packages/Arcus.BackgroundJobs.Databricks/). This package contains all the necessary capabilities to query the Databricks cluster so we can retrieve the result of the completed jobs. 
```

> We'll use an Azure Function to create a timer that will periodically query this Databricks cluster.

Reviewer: ```suggestion
We'll use an Azure Function that will periodically query our Databricks cluster by using a timer trigger.
```

> The job outcome itself will be written as an Application Insights Metric so we can use it later as a source of an alert.

Reviewer: ```suggestion
The job outcome itself will be written as an Azure Application Insights Metric so we can use it later as a source of an alert.
```

            {
                await provider.MeasureJobOutcomesAsync(metricName, startOfWindow, endOfWindow);
            }
        }
    }
}
```

The `DatabricksInfoProvider` is an Arcus type that allows you to either measure the job outcomes directly, or query them so you can log them yourself. For more information on this type, see [the official docs](https://background-jobs.arcus-azure.net/features/databricks/gain-insights).

Note that we specify a time window in which we want the completed jobs to exist. This is now set on the start and end of each run of the Azure Function time trigger.

## Application Insights metrics and alerts

All the previous setup  will result in log metrics like the following:
```
Metric Databricks Job Completed: 1 at 12/10/2020 08:50:01 +00:00 (Context: [Run Id, 77], [Job Id, 9], [Job Name, My Databricks Job], [Outcome, success])
```

If you want them to be in Application Insights, please use [our own Application Insights Serilog sink](https://observability.arcus-azure.net/features/sinks/azure-application-insights) so the written log message is discovered as metrics and not traces.

The metrics will show up as custom metrics:

> ![Custom metrics](./images/custom-metrics.png)

Reviewer: You could also use the metric explorer to show the graph; if you want

Author: Yeah, but the metric values are always `1` in this case; not sure if that would add much of added-value. Showing the actual logs with the job outcome may be more appropriate here.

Reviewer: Yes because you could sum() the values, but it's ok

The Databricks job outcome is available as custom dimension in the metric:

![Custom dimensions](./images/custom-dimensions.png)

If you want, you can create an alert on this metric by going to **Application Insights > Monitoring > Alerts > New alert rule**.

> ![New alert rule](./images/new-alert-rule.png)

Reviewer: I would leave this one out since it doesn't have that much value

Author: Done this, resolving.

There, in the alert condition, the Databricks metric will be available to you.

![Alert condition](./images/alert-condition.png)

# Conclusion

> In this post we have looked on how we can monitor Databricks jobs with the new **Arcus BackgroundJobs** library. We're not inventing anything new here, only integreting existing services: Databricks job outcomes to Azure Function to Application Insights.

Reviewer: ```suggestion
In this post, we have looked at how we can monitor Databricks jobs with the new **Arcus Background Jobs** library. We're not inventing anything new here, only integrating existing services - Querying Databricks job outcomes in an Azure Function and flowing the information to Azure Application Insights.
```

You'll be happy to know that we have created an project template for this with already the necessary Azure Function and packages embedded.

> This template will be available starting from v0.5 in the [Arcus Templates](https://templates.arcus-azure.net/) library.

Author: Yeah okay, but it doesn't say it's available directly, only starting from the next version.

> This template will be available starting from v0.5 in the [Arcus Templates](https://templates.arcus-azure.net/) library.

Reviewer: Oh I've missed that, sorry
